js_modules/dagster-ui/packages/ui-core/src/hooks/useDangerousRenderEffect.tsx

```diff
@@ -0,0 +1,22 @@
+import {DependencyList, EffectCallback, useRef} from 'react';
+/**
+ * DO NOT CALL `setState` from within this effect or from any function called by this effect.
+ * THIS IS NOT INTENDED TO TRIGGER REACT UPDATES.
+ * This is intended for telemetry and even then this is not Suspense safe
+ * because it mutates refs during render and Suspense blows away all component state (including refs).
+ * Use with extreme care.
+ */
+
+export function useDangerousRenderEffect(effect: EffectCallback, deps: DependencyList) {
+  const initialRender = useRef(true);
+  const depsRef = useRef(deps);
+
+  const dependenciesChanged = deps.some((dep, index) => dep !== depsRef.current[index]);
+
+  // Run effect if it's the first render or if dependencies have changed
+  if (initialRender.current || dependenciesChanged) {
+    depsRef.current = deps;
+    initialRender.current = false;
+    effect();
+  }
+}
```

Comment:
Basically react components are supposed to be pure. The way to do this properly is with some global state via Suspense cache

Reply:
This looks good to me, but I think we do this in other places more directly by abusing useMemo?

Because useMemo runs and returns immediately, I think this is the same as:

```
useMemo(() => {
    effect();
}, [deps);
```

Reply:
Yeah that would technically be the same thing but I definitely don't want to encourage that anti-pattern. I'll replace the implementation with that though